### PR TITLE
v-ramp render on visible

### PR DESCRIPTION
**Pull request by gerhean**
Possible solution to [#1383]

Original (18s):
https://nus-cs2103-ay2021s1.github.io/tp-dashboard/

Demo (4s):
https://gerhean.github.io/tp-dashboard/

```
v-ramp is clearly very expensive to render.  The user
is unlikely to scroll through the entire list of repos 
rendered, making it a waste to render everything.

Let's render v-ramp only when visible, hence
massively reducing initial load time.
```

### CODE REVIEW

**Comment by HCY123902** [comment]
Can you show the time taken to load this particular report with and without this change to draw a comparison?

**Comment by gerhean** [comment]
I think it is better for you to test it yourself as there are some visual differences.

**Review by dcshzj — APPROVED** [review]
LGTM. RAM usage looks significantly lesser on first load.

File: frontend/src/index.pug [hunk 1]
```diff
@@ -25,6 +25,7 @@ html
     script(src="https://cdn.jsdelivr.net/npm/vue-loading-overlay@3")
     script(src="https://cdn.jsdelivr.net/npm/vue@2.5/dist/vue.js")
     script(src="https://unpkg.com/vuex@3.1.3/dist/vuex.js")
+    script(src="https://unpkg.com/vue-observe-visibility/dist/vue-observe-visibility.min.js")
```
Comment (dcshzj) [inline frontend/src/index.pug:28]: One thing, can you specify the exact version to use?

**Review by dcshzj — CHANGES_REQUESTED** [review]
As above.

**Review by dcshzj — APPROVED** [review]
LGTM!

**Comment by gerhean** [comment]
Actually I was wondering about the purpose of specifying the rampSize in the state previously, if anyone can answer.

**Comment by fzdy1914** [comment]
@damithc Do you think the render on visible OK to you?

**Comment by damithc** [comment]
> @damithc Do you think the render on visible OK to you?

Seems significantly faster. And seems memory usage is less too. Nice. 
1. But this comparison may not be fair as the first dashboard is using a much older version of RepoSense. Can we compare with the current master?
2. Are there any known downsides?

**Comment by gerhean** [comment]
One obvious downside:
image

If you scroll too fast the ramps take about 1s to actually appear.

**Comment by gerhean** [comment]
Actually when you load every single ramp, the filtering becomes slow again.

**Comment by damithc** [comment]
> If you scroll too fast the ramps take about 1s to actually appear (aesthetic factor).

Might be OK if they appear fast enough, without misleading the user into thinking there are no ramps. The alternative is to show some placeholder first e.g., a `loading ...` but not sure if it will make the page slow again. 
BTW, it seems to take longer (even 2-3 seconds at times) for the ramps to appear in my home computer.

**Comment by fzdy1914** [comment]
@gerhean Can we do a mix-up? We first only render the ramps that are visible and display them to the users. Then, we start to render all other ramps.

**Comment by damithc** [comment]
> @gerhean Can we do a mix-up? We first only render the ramps that are visible and display them to the users. Then, we start to render all other ramps regardless of visible or not.

Yes, that feels like best of both worlds.

**Comment by fzdy1914** [comment]
@gerhean Any update?

**Comment by gerhean** [comment]
There's still a few issues to solve like the loading spinner not displaying properly.

**Pull request closed without merging**